### PR TITLE
chore: optimize CB batch time by running tasks in 2 waves

### PR DESCRIPTION
**Pull request by goldbez**
#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codebuild_specs/e2e_workflow_base.yml [hunk 1]
```diff
@@ -32,68 +32,41 @@ batch:
       buildspec: codebuild_specs/test.yml
       env:
         compute-type: BUILD_GENERAL1_LARGE
```
Comment (awsluja) [inline codebuild_specs/e2e_workflow_base.yml:34]: can you remove these depend-on relationships in the e2e_workflow.yml file as well? so we get the same benefits when running locally

**Review by awsluja — APPROVED** [review]

File: codebuild_specs/mock_e2e_tests.yml [hunk 2]
```diff
@@ -7,4 +7,5 @@ env:
 phases:
   build:
     commands:
+      - source shared-scripts.sh && _waitForJobs build_linux
       - source ./shared-scripts.sh && _mockE2ETests
```
Comment (edwardfoyle) [inline codebuild_specs/mock_e2e_tests.yml:11]: we only need to source shared-scripts once right?

File: scripts/wait-for-all-codebuild.ts [hunk 3]
```diff
@@ -29,8 +29,13 @@ const main = async () => {
...
   const jobsDependedOnFilepath = process.argv[3];
   const codeBuildProjectName = process.argv[4];
-  const jobsDependedOnRaw = fs.readFileSync(jobsDependedOnFilepath, 'utf8');
-  const jobsDependedOn = JSON.parse(jobsDependedOnRaw);
+  let jobsDependedOn: string[];
+  try {
+    const jobsDependedOnRaw = fs.readFileSync(jobsDependedOnFilepath, 'utf8');
+    jobsDependedOn = JSON.parse(jobsDependedOnRaw);
+  } catch {
+    jobsDependedOn = [jobsDependedOnFilepath];
```
Comment (edwardfoyle) [inline scripts/wait-for-all-codebuild.ts:37]: in what cases would this condition hit?

File: codebuild_specs/mock_e2e_tests.yml (hunk 2 above)
Comment (awsluja) [reply codebuild_specs/mock_e2e_tests.yml:11]: Unfortunately we've found that codebuild does not behave this wave; Each command has its own source/context with codebuild within the "commands" section

File: scripts/wait-for-all-codebuild.ts (hunk 3 above)
Comment (awsluja) [reply scripts/wait-for-all-codebuild.ts:37]: If jobsDependedOnFilepath is the name of a job, a file is not needed. We take jobsDependedOnFilepath and create an array with it as a single element. 
For example, this allows us to call:
_waitForJobs <jobname>
or
_waitForJobs <filepath with job names list>

Comment (edwardfoyle) [reply scripts/wait-for-all-codebuild.ts:37]: We should make a backlog item to revisit this. There's a lot of implicit nuance here like it seems that `jobsDependedOnFilepath` is not always a file path or what happens if the JSON content is malformed such that `JSON.parse()` fails but was intended to be parsed?

**Review by edwardfoyle — APPROVED** [review]

**Pull request merged**